File: internal/infra/ffprobe.go
package infra

import (
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"

	"github.com/freesiapro/resize-to-telegram-sticker/internal/domain"
)

type FFprobeRunner struct {
	Path string
}

type probeJSON struct {
	Streams []struct {
		CodecType string `json:"codec_type"`
		Width     int    `json:"width"`
		Height    int    `json:"height"`
		FrameRate string `json:"r_frame_rate"`
		CodecName string `json:"codec_name"`
		Duration  string `json:"duration"`
	} `json:"streams"`
	Format struct {
		FormatName string `json:"format_name"`
		Duration   string `json:"duration"`
		BitRate    string `json:"bit_rate"`
	} `json:"format"`
}

func (r FFprobeRunner) Probe(ctx context.Context, path string) (domain.MediaInfo, error) {
	bin := r.Path
	if bin == "" {
		bin = "ffprobe"
	}
	args := []string{
		"-v", "error",
		"-show_entries", "stream=codec_type,width,height,r_frame_rate,codec_name,duration",
		"-show_entries", "format=duration,format_name",
		"-of", "json",
		path,
	}
	out, err := exec.CommandContext(ctx, bin, args...).Output()
	if err != nil {
		return domain.MediaInfo{}, fmt.Errorf("ffprobe failed: %w", err)
	}
	return parseProbeJSON(out)
}

func parseProbeJSON(data []byte) (domain.MediaInfo, error) {
	var p probeJSON
	if err := json.Unmarshal(data, &p); err != nil {
		return domain.MediaInfo{}, err
	}

	info := domain.MediaInfo{}
	for _, s := range p.Streams {
		if s.CodecType == "audio" {
			info.HasAudio = true
		}
		if s.CodecType == "video" {
			info.Width = s.Width
			info.Height = s.Height
			info.FPS = parseFrameRate(s.FrameRate)
			info.CodecName = s.CodecName
			if info.DurationSeconds == 0 {
				info.DurationSeconds = parseDuration(s.Duration)
			}
		}
	}
	if info.DurationSeconds == 0 {
		info.DurationSeconds = parseDuration(p.Format.Duration)
	}
	info.FormatName = p.Format.FormatName
	info.BitrateBps = parseBitRate(p.Format.BitRate)

	return info, nil
}

func parseFrameRate(v string) float64 {
	parts := strings.Split(v, "/")
	if len(parts) != 2 {
		return 0
	}
	num, _ := strconv.ParseFloat(parts[0], 64)
	den, _ := strconv.ParseFloat(parts[1], 64)
	if den == 0 {
		return 0
	}
	return num / den
}

func parseDuration(v string) float64 {
	f, _ := strconv.ParseFloat(v, 64)
	return f
}

func parseBitRate(v string) int64 {
	if v == "" {
		return 0
	}
	rate, err := strconv.ParseInt(v, 10, 64)
	if err == nil {
		return rate
	}
	f, err := strconv.ParseFloat(v, 64)
	if err != nil {
		return 0
	}
	return int64(f)
}

File: python/infra/ffprobe.py
from __future__ import annotations

import json
import subprocess
from dataclasses import dataclass

from core.media import MediaInfo


@dataclass
class FFprobeRunner:
    path: str = "ffprobe"

    def probe(self, media_path: str) -> MediaInfo:
        args = [
            self.path,
            "-v",
            "error",
            "-show_entries",
            "stream=codec_type,width,height,r_frame_rate,codec_name,duration",
            "-show_entries",
            "format=duration,format_name,bit_rate",
            "-of",
            "json",
            media_path,
        ]
        proc = subprocess.run(args, capture_output=True, text=True, check=False)
        if proc.returncode != 0:
            stderr = proc.stderr.strip()
            raise RuntimeError(f"ffprobe failed: {stderr}")
        return parse_probe_json(proc.stdout)


def parse_probe_json(payload: str) -> MediaInfo:
    data = json.loads(payload)
    info = MediaInfo()

    for stream in data.get("streams", []):
        codec_type = stream.get("codec_type")
        if codec_type == "audio":
            info.has_audio = True
        if codec_type == "video":
            info.width = int(stream.get("width") or 0)
            info.height = int(stream.get("height") or 0)
            info.fps = parse_frame_rate(str(stream.get("r_frame_rate") or "0/0"))
            info.codec_name = str(stream.get("codec_name") or "")
            if info.duration_seconds == 0:
                info.duration_seconds = parse_duration(str(stream.get("duration") or "0"))

    fmt = data.get("format", {})
    if info.duration_seconds == 0:
        info.duration_seconds = parse_duration(str(fmt.get("duration") or "0"))
    info.format_name = str(fmt.get("format_name") or "")
    info.bitrate_bps = parse_bitrate(str(fmt.get("bit_rate") or "0"))

    return info


def parse_frame_rate(value: str) -> float:
    parts = value.split("/")
    if len(parts) != 2:
        return 0.0
    try:
        num = float(parts[0])
        den = float(parts[1])
    except ValueError:
        return 0.0
    if den == 0:
        return 0.0
    return num / den


def parse_duration(value: str) -> float:
    try:
        return float(value)
    except ValueError:
        return 0.0


def parse_bitrate(value: str) -> int:
    if not value:
        return 0
    try:
        return int(value)
    except ValueError:
        try:
            return int(float(value))
        except ValueError:
            return 0
